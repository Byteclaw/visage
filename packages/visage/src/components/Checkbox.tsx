import { useStaticCallbackCreator } from '@byteclaw/use-static-callback';
import {
  ExtractVisageComponentProps,
  VisageComponent,
  markAsVisageComponent,
} from '@byteclaw/visage-core';
import React, {
  ReactNode,
  forwardRef,
  memo,
  Ref,
  useState,
  FocusEventHandler,
} from 'react';
import { createComponent } from '../core';
import {
  visuallyHiddenStyles,
  visuallyHiddenBooleanVariant,
  disabledControlBooleanVariant,
  disabledControlStyles,
  createControlFocusShadow,
} from './shared';
import { Flex } from './Flex';
import { Svg } from './Svg';
import { booleanVariant, booleanVariantStyles } from '../variants';
import { useComposedCallbackCreator, useHandlerRef } from '../hooks';
import { wrapToggleOnChangeHandler } from './events';

const CheckboxControl = createComponent('input', {
  displayName: 'CheckboxControl',
  styles: {
    ...visuallyHiddenStyles,
  },
});

const CheckboxLabel = createComponent('label', {
  displayName: 'CheckboxLabel',
  styles: props => ({
    alignItems: 'flex-start',
    cursor: 'pointer',
    display: 'flex',
    flex: '1',
    fontSize: 'inherit',
    lineHeight: 'inherit',
    m: 0,
    outline: 'none',
    p: 0,
    position: 'relative',
    userSelect: 'none',
    ...(props.disabled ? disabledControlStyles : {}),
  }),
  variants: [disabledControlBooleanVariant],
});

const CheckboxLabelText = createComponent('span', {
  displayName: 'CheckboxLabelText',
  styles: props => ({
    fontSize: 'inherit',
    lineHeight: 'inherit',
    ...(props.hidden ? visuallyHiddenStyles : {}),
  }),
  variants: [visuallyHiddenBooleanVariant],
});

const CheckboxToggler = createComponent(Flex, {
  displayName: 'CheckboxToggler',
  defaultProps: {
    'aria-hidden': true,
    children: (
      <Svg
        viewBox="0 0 24 24"
        styles={{
          width: '1em',
          height: '1em',
          stroke: 'color(shades tint(10%))',
          strokeWidth: '2px',
          fill: 'none',
        }}
      >
        <polyline points="20 6 9 17 4 12" />
      </Svg>
    ),
  },
  styles: {
    backgroundColor: 'color(shades tint(10%))',
    alignSelf: 'center',
    transition: 'all 150ms',
    borderColor: 'accent',
    borderRadius: '3px',
    borderStyle: 'solid',
    borderWidth: '2px',
    mr: 1,
    ...booleanVariantStyles('checked', {
      on: {
        backgroundColor: 'primary',
        color: 'primaryText',
        borderColor: 'primary',
        '& > svg': {
          visibility: 'visible',
        },
      },
      off: {
        '& > svg': {
          visibility: 'hidden',
        },
      },
    }),
    ...booleanVariantStyles('invalid', {
      on: {
        borderColor: 'danger',
        ...booleanVariantStyles('focused', {
          on: {
            boxShadow: createControlFocusShadow('danger'),
          },
        }),
      },
      off: booleanVariantStyles('focused', {
        on: {
          boxShadow: createControlFocusShadow(),
        },
      }),
    }),
  },
  variants: [
    booleanVariant('checked', true),
    booleanVariant('disabled', true),
    booleanVariant('invalid', true),
    booleanVariant('focused', true),
    booleanVariant('readOnly', true),
  ],
});

interface TogglerProps {
  checked: boolean;
  disabled?: boolean;
  focused: boolean;
  invalid?: boolean;
  readOnly?: boolean;
}

const DefaultCheckboxToggler = memo((props: TogglerProps) => {
  return <CheckboxToggler {...props} />;
});

interface CheckboxProps
  extends ExtractVisageComponentProps<typeof CheckboxControl> {
  /**
   * Hides label visually
   */
  hiddenLabel?: boolean;
  /**
   * Sets the checkbox aria-invalid to true and applies visual style
   */
  invalid?: boolean;
  /**
   * Label is required to be WAI-ARIA compliant
   */
  label: ReactNode;
  /**
   * Passes props to the label
   */
  labelProps?: ExtractVisageComponentProps<typeof CheckboxLabel>;
  /**
   * Passes props to the label text
   */
  labelTextProps?: ExtractVisageComponentProps<typeof CheckboxLabelText>;
  ref?: React.RefObject<HTMLInputElement>;
  /**
   * Toggler componet
   */
  toggler?: React.ComponentType<TogglerProps>;
}

export const Checkbox: VisageComponent<CheckboxProps> = memo(
  forwardRef(function Checkbox(
    {
      $$variants,
      checked,
      defaultChecked,
      disabled,
      hiddenLabel = false,
      invalid,
      label,
      labelProps,
      labelTextProps,
      onBlur,
      onFocus,
      onChange,
      parentStyles,
      readOnly,
      styles,
      toggler: Toggler = DefaultCheckboxToggler,
      ...rest
    }: CheckboxProps,
    ref: Ref<HTMLInputElement>,
  ) {
    const [focused, setFocused] = useState(false);
    const [innerChecked, setInnerChecked] = useState(
      checked ?? defaultChecked ?? false,
    );
    const onInnerBlur = useHandlerRef(() => setFocused(false));
    const onInnerFocus = useHandlerRef(() => setFocused(true));
    const onBlurHandler = useComposedCallbackCreator<FocusEventHandler>(
      onBlur,
      onInnerBlur,
    );
    const onFocusHandler = useComposedCallbackCreator<FocusEventHandler>(
      onFocus,
      onInnerFocus,
    );
    const onChangeHandler = useStaticCallbackCreator(
      wrapToggleOnChangeHandler,
      [readOnly, onChange, setInnerChecked],
    );
    // if onChange is provided then the component is controlled
    const isChecked = onChange ? checked ?? false : innerChecked;

    return (
      <CheckboxLabel
        {...labelProps}
        disabled={disabled}
        parentStyles={parentStyles}
        styles={styles}
        $$variants={$$variants}
      >
        <CheckboxControl
          {...rest}
          checked={isChecked}
          aria-invalid={invalid}
          disabled={disabled}
          onBlur={onBlurHandler}
          onChange={onChangeHandler}
          onFocus={onFocusHandler}
          ref={ref}
          readOnly={readOnly}
          type="checkbox"
        />
        <Toggler
          checked={isChecked}
          disabled={disabled}
          focused={focused}
          invalid={invalid}
          readOnly={readOnly}
        />
        &#8203; {/* fixes height if label is hidden */}
        <CheckboxLabelText {...labelTextProps} hidden={hiddenLabel}>
          {label}
        </CheckboxLabelText>
      </CheckboxLabel>
    );
  }),
);

markAsVisageComponent(Checkbox);
