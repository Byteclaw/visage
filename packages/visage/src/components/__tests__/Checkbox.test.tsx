/* eslint-disable no-irregular-whitespace */
import { fireEvent } from '@testing-library/react';
import React from 'react';
import { render } from './render';
import { Checkbox } from '../Checkbox';

describe('Checkbox', () => {
  it('respects disabled prop', () => {
    const handler = jest.fn();
    const { getByTestId } = render(
      <Checkbox
        data-testid="checkbox"
        disabled
        onClick={handler}
        onKeyDown={handler}
        label=""
        labelProps={{ 'data-testid': 'label' }}
      />,
    );

    expect(getByTestId('checkbox')).toHaveAttribute('disabled');
    expect(getByTestId('label')).toHaveAttribute('disabled');

    fireEvent.click(getByTestId('checkbox'));
    // we don't fire a keyDown because it isn't possible to focus disabled checkbox in browser

    expect(handler).not.toHaveBeenCalled();
  });

  it('respects hiddenLabel prop', () => {
    const { getByTestId } = render(
      <Checkbox
        data-testid="checkbox"
        hiddenLabel
        label="Label"
        labelProps={{ 'data-testid': 'label' }}
        labelTextProps={{ 'data-testid': 'label-text' }}
      />,
    );

    expect(getByTestId('label-text')).toHaveAttribute('data-hidden', 'true');
  });

  it('respects invalid prop', () => {
    const props = {
      'data-testid': 'checkbox',
      label: 'Label',
      labelProps: { 'data-test-id': 'label' },
      labelTextProps: { 'date-testid': 'label-text' },
    };
    const { getByTestId, rerender } = render(<Checkbox {...props} />);

    expect(getByTestId('checkbox')).not.toHaveAttribute('aria-invalid');

    rerender(<Checkbox {...props} invalid />);

    expect(getByTestId('checkbox')).toHaveAttribute('aria-invalid', 'true');
  });

  it('respects readOnly prop', () => {
    const handler = jest.fn();
    const { getByTestId } = render(
      <Checkbox
        data-testid="checkbox"
        readOnly
        onClick={handler}
        onKeyDown={handler}
        label=""
        labelProps={{ 'data-testid': 'label' }}
      />,
    );

    // read only checkbox can be focused, clicked, etc but can't be changed
    expect(getByTestId('checkbox')).toHaveAttribute('readOnly');

    fireEvent.click(getByTestId('checkbox'));
    fireEvent.keyDown(getByTestId('checkbox'), { key: ' ' });

    expect(handler).toHaveBeenCalledTimes(2);
  });

  it('renders correctly', () => {
    const { asFragment } = render(<Checkbox label="Required label" />);

    expect(asFragment()).toMatchInlineSnapshot(`
      <DocumentFragment>
        .emotion-0 {
        -webkit-box-pack: center;
        -webkit-justify-content: center;
        -ms-flex-pack: center;
        justify-content: center;
        display: -webkit-box;
        display: -webkit-flex;
        display: -ms-flexbox;
        display: flex;
        position: fixed;
        bottom: 0;
        left: 0;
        right: 0;
        z-index: 600;
      }

      <div
          aria-live="polite"
          class="emotion-0"
          data-toastmanager="true"
        />
        .emotion-0 {
        border: 0;
        -webkit-clip: rect(0,0,0,0);
        clip: rect(0,0,0,0);
        height: 1px;
        overflow: hidden;
        margin: -1px;
        padding: 0px;
        position: absolute;
        white-space: nowrap;
        width: 1px;
      }

      .emotion-0:focus + div {
        box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px lightAccent,inset 0 0 200px rgba(68,68,68,0.1);
      }

      .emotion-0 + div {
        background-color: textInput;
      }

      .emotion-0:checked + div {
        background-color: light-blue;
        color: white;
        border-color: light-blue;
      }

      .emotion-0[aria-invalid="true"] + div {
        border-color: danger;
      }

      .emotion-0 + div > svg {
        visibility: hidden;
      }

      .emotion-0:checked + div > svg {
        visibility: visible;
      }

      .emotion-2 {
        display: -webkit-box;
        display: -webkit-flex;
        display: -ms-flexbox;
        display: flex;
        -webkit-align-self: center;
        -ms-flex-item-align: center;
        align-self: center;
        -webkit-transition: all 150ms;
        transition: all 150ms;
        border-color: lightAccent;
        border-radius: 3px;
        border-style: solid;
        border-width: 2px;
        margin-right: 8px;
      }

      .emotion-1 {
        width: 1em;
        height: 1em;
        stroke: textInput;
        stroke-width: 2px;
        fill: none;
      }

      .emotion-3 {
        font-size: inherit;
        line-height: inherit;
      }

      .emotion-4 {
        -webkit-align-items: flex-start;
        -webkit-box-align: flex-start;
        -ms-flex-align: flex-start;
        align-items: flex-start;
        cursor: pointer;
        display: -webkit-box;
        display: -webkit-flex;
        display: -ms-flexbox;
        display: flex;
        -webkit-flex: 1;
        -ms-flex: 1;
        flex: 1;
        font-size: inherit;
        line-height: inherit;
        margin: 0px;
        outline: none;
        padding: 0px;
        position: relative;
        -webkit-user-select: none;
        -moz-user-select: none;
        -ms-user-select: none;
        user-select: none;
      }

      <label
          class="emotion-4"
          data-disabled="false"
        >
          <input
            class="emotion-0"
            type="checkbox"
          />
          <div
            aria-hidden="true"
            class="emotion-2"
          >
            <svg
              class="emotion-1"
              viewBox="0 0 24 24"
            >
              <polyline
                points="20 6 9 17 4 12"
              />
            </svg>
          </div>
          â€‹ 
          <span
            class="emotion-3"
            data-hidden="false"
          >
            Required label
          </span>
        </label>
      </DocumentFragment>
    `);
  });
});
