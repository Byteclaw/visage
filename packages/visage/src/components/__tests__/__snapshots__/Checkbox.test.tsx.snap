// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox is extendable by CheckboxControl face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: rgb(209,209,209);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  border-color: rgb(0,0,255);
  border-radius: 3px;
  border-style: solid;
  border-width: 2px;
  margin-right: 8px;
}

.emotion-2[data-checked="true"] {
  background-color: light-blue;
  color: rgb(255,255,255);
  border-color: light-blue;
}

.emotion-2[data-checked="true"] > svg {
  visibility: visible;
}

.emotion-2[data-checked="false"] > svg {
  visibility: hidden;
}

.emotion-2[data-invalid="true"] {
  border-color: rgb(255,0,0);
}

.emotion-2[data-invalid="true"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(255,0,0);
}

.emotion-2[data-invalid="false"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255);
}

.emotion-1 {
  width: 1em;
  height: 1em;
  stroke: rgb(209,209,209);
  stroke-width: 2px;
  fill: none;
}

.emotion-3 {
  font-size: inherit;
  line-height: inherit;
}

.emotion-4 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  line-height: inherit;
  margin: 0px;
  outline: none;
  padding: 0px;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-0 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
  border-radius: 10px;
}

<label
    class="emotion-4"
    data-disabled="false"
  >
    <input
      class="emotion-0"
      type="checkbox"
    />
    <div
      aria-hidden="true"
      class="emotion-2"
      data-checked="false"
      data-disabled="false"
      data-focused="false"
      data-invalid="false"
      data-readonly="false"
    >
      <svg
        class="emotion-1"
        viewBox="0 0 24 24"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </div>
    ​ 
    <span
      class="emotion-3"
      data-hidden="false"
    >
      test
    </span>
  </label>
</DocumentFragment>
`;

exports[`Checkbox is extendable by CheckboxLabel face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-0 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: rgb(209,209,209);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  border-color: rgb(0,0,255);
  border-radius: 3px;
  border-style: solid;
  border-width: 2px;
  margin-right: 8px;
}

.emotion-2[data-checked="true"] {
  background-color: light-blue;
  color: rgb(255,255,255);
  border-color: light-blue;
}

.emotion-2[data-checked="true"] > svg {
  visibility: visible;
}

.emotion-2[data-checked="false"] > svg {
  visibility: hidden;
}

.emotion-2[data-invalid="true"] {
  border-color: rgb(255,0,0);
}

.emotion-2[data-invalid="true"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(255,0,0);
}

.emotion-2[data-invalid="false"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255);
}

.emotion-1 {
  width: 1em;
  height: 1em;
  stroke: rgb(209,209,209);
  stroke-width: 2px;
  fill: none;
}

.emotion-3 {
  font-size: inherit;
  line-height: inherit;
}

.emotion-4 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  line-height: inherit;
  margin: 0px;
  outline: none;
  padding: 0px;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 10px;
}

<label
    class="emotion-4"
    data-disabled="false"
  >
    <input
      class="emotion-0"
      type="checkbox"
    />
    <div
      aria-hidden="true"
      class="emotion-2"
      data-checked="false"
      data-disabled="false"
      data-focused="false"
      data-invalid="false"
      data-readonly="false"
    >
      <svg
        class="emotion-1"
        viewBox="0 0 24 24"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </div>
    ​ 
    <span
      class="emotion-3"
      data-hidden="false"
    >
      test
    </span>
  </label>
</DocumentFragment>
`;

exports[`Checkbox is extendable by CheckboxLabelText face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-0 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: rgb(209,209,209);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  border-color: rgb(0,0,255);
  border-radius: 3px;
  border-style: solid;
  border-width: 2px;
  margin-right: 8px;
}

.emotion-2[data-checked="true"] {
  background-color: light-blue;
  color: rgb(255,255,255);
  border-color: light-blue;
}

.emotion-2[data-checked="true"] > svg {
  visibility: visible;
}

.emotion-2[data-checked="false"] > svg {
  visibility: hidden;
}

.emotion-2[data-invalid="true"] {
  border-color: rgb(255,0,0);
}

.emotion-2[data-invalid="true"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(255,0,0);
}

.emotion-2[data-invalid="false"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255);
}

.emotion-1 {
  width: 1em;
  height: 1em;
  stroke: rgb(209,209,209);
  stroke-width: 2px;
  fill: none;
}

.emotion-4 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  line-height: inherit;
  margin: 0px;
  outline: none;
  padding: 0px;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-3 {
  font-size: inherit;
  line-height: inherit;
  border-radius: 10px;
}

<label
    class="emotion-4"
    data-disabled="false"
  >
    <input
      class="emotion-0"
      type="checkbox"
    />
    <div
      aria-hidden="true"
      class="emotion-2"
      data-checked="false"
      data-disabled="false"
      data-focused="false"
      data-invalid="false"
      data-readonly="false"
    >
      <svg
        class="emotion-1"
        viewBox="0 0 24 24"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </div>
    ​ 
    <span
      class="emotion-3"
      data-hidden="false"
    >
      test
    </span>
  </label>
</DocumentFragment>
`;

exports[`Checkbox is extendable by CheckboxToggler face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-0 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-1 {
  width: 1em;
  height: 1em;
  stroke: rgb(209,209,209);
  stroke-width: 2px;
  fill: none;
}

.emotion-3 {
  font-size: inherit;
  line-height: inherit;
}

.emotion-4 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  line-height: inherit;
  margin: 0px;
  outline: none;
  padding: 0px;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: rgb(209,209,209);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  border-color: rgb(0,0,255);
  border-radius: 10px;
  border-style: solid;
  border-width: 2px;
  margin-right: 8px;
}

.emotion-2[data-checked="true"] {
  background-color: light-blue;
  color: rgb(255,255,255);
  border-color: light-blue;
}

.emotion-2[data-checked="true"] > svg {
  visibility: visible;
}

.emotion-2[data-checked="false"] > svg {
  visibility: hidden;
}

.emotion-2[data-invalid="true"] {
  border-color: rgb(255,0,0);
}

.emotion-2[data-invalid="true"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(255,0,0);
}

.emotion-2[data-invalid="false"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255);
}

<label
    class="emotion-4"
    data-disabled="false"
  >
    <input
      class="emotion-0"
      type="checkbox"
    />
    <div
      aria-hidden="true"
      class="emotion-2"
      data-checked="false"
      data-disabled="false"
      data-focused="false"
      data-invalid="false"
      data-readonly="false"
    >
      <svg
        class="emotion-1"
        viewBox="0 0 24 24"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </div>
    ​ 
    <span
      class="emotion-3"
      data-hidden="false"
    >
      test
    </span>
  </label>
</DocumentFragment>
`;

exports[`Checkbox is extendable by styles 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-0 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: rgb(209,209,209);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  border-color: rgb(0,0,255);
  border-radius: 3px;
  border-style: solid;
  border-width: 2px;
  margin-right: 8px;
}

.emotion-2[data-checked="true"] {
  background-color: light-blue;
  color: rgb(255,255,255);
  border-color: light-blue;
}

.emotion-2[data-checked="true"] > svg {
  visibility: visible;
}

.emotion-2[data-checked="false"] > svg {
  visibility: hidden;
}

.emotion-2[data-invalid="true"] {
  border-color: rgb(255,0,0);
}

.emotion-2[data-invalid="true"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(255,0,0);
}

.emotion-2[data-invalid="false"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255);
}

.emotion-1 {
  width: 1em;
  height: 1em;
  stroke: rgb(209,209,209);
  stroke-width: 2px;
  fill: none;
}

.emotion-3 {
  font-size: inherit;
  line-height: inherit;
}

.emotion-4 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  line-height: inherit;
  margin: 0px;
  outline: none;
  padding: 0px;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: rgb(255,192,203);
}

<label
    class="emotion-4"
    data-disabled="false"
  >
    <input
      class="emotion-0"
      type="checkbox"
    />
    <div
      aria-hidden="true"
      class="emotion-2"
      data-checked="false"
      data-disabled="false"
      data-focused="false"
      data-invalid="false"
      data-readonly="false"
    >
      <svg
        class="emotion-1"
        viewBox="0 0 24 24"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </div>
    ​ 
    <span
      class="emotion-3"
      data-hidden="false"
    >
      test
    </span>
  </label>
</DocumentFragment>
`;

exports[`Checkbox renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-0 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: rgb(209,209,209);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  border-color: rgb(0,0,255);
  border-radius: 3px;
  border-style: solid;
  border-width: 2px;
  margin-right: 8px;
}

.emotion-2[data-checked="true"] {
  background-color: light-blue;
  color: rgb(255,255,255);
  border-color: light-blue;
}

.emotion-2[data-checked="true"] > svg {
  visibility: visible;
}

.emotion-2[data-checked="false"] > svg {
  visibility: hidden;
}

.emotion-2[data-invalid="true"] {
  border-color: rgb(255,0,0);
}

.emotion-2[data-invalid="true"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(255,0,0);
}

.emotion-2[data-invalid="false"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255);
}

.emotion-1 {
  width: 1em;
  height: 1em;
  stroke: rgb(209,209,209);
  stroke-width: 2px;
  fill: none;
}

.emotion-3 {
  font-size: inherit;
  line-height: inherit;
}

.emotion-4 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  line-height: inherit;
  margin: 0px;
  outline: none;
  padding: 0px;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

<label
    class="emotion-4"
    data-disabled="false"
  >
    <input
      class="emotion-0"
      type="checkbox"
    />
    <div
      aria-hidden="true"
      class="emotion-2"
      data-checked="false"
      data-disabled="false"
      data-focused="false"
      data-invalid="false"
      data-readonly="false"
    >
      <svg
        class="emotion-1"
        viewBox="0 0 24 24"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </div>
    ​ 
    <span
      class="emotion-3"
      data-hidden="false"
    >
      Required label
    </span>
  </label>
</DocumentFragment>
`;
